src/NuGet.Clients/NuGet.SolutionRestoreManager/VsSolutionRestoreService.cs

```diff
@@ -309,25 +309,132 @@ private static string GetProjectOutputPath(string projectDirectory, string msbui
         }
+
+        private async Task<ProjectNames> GetProjectNamesAsync(string projectUniqueName, CancellationToken token)
+        {
+            if (!_projectSystemCache.TryGetProjectNames(projectUniqueName, out ProjectNames projectNames))
+            {
+                IVsSolution2 vsSolution2 = await _vsSolution2.GetValueAsync(token);
+                projectNames = await ProjectNames.FromIVsSolution2(projectUniqueName, vsSolution2, token);
+            }
+
+            return projectNames;
+        }
+
+        /// <summary>
+        /// Populates restore info sources.
+        /// </summary>
+        private async Task PopulateRestoreInfoSourcesAsync()
+        {
+            List<(IVsProjectRestoreInfoSource, CancellationToken, TaskCompletionSource<bool>)> failedInits = new List<(IVsProjectRestoreInfoSource, CancellationToken, TaskCompletionSource<bool>)>();
+            while (_projectRestoreInfoSources.TryDequeue(out var source))
+            {
+                IVsProjectRestoreInfoSource projectRestoreInfoSource = source.Item1;
+                string projectUniqueName = source.Item1.Name;
+                CancellationToken token = source.Item2;
+                TaskCompletionSource<bool> taskCompletionSource = source.Item3;
+
+                if (!_projectSystemCache.TryGetProjectNames(projectUniqueName, out ProjectNames projectNames))
+                {
+                    IVsSolution2 vsSolution2 = await _vsSolution2.GetValueAsync(token);
+
+                    try
+                    {
+                        token.ThrowIfCancellationRequested();
+                        projectNames = await ProjectNames.FromIVsSolution2(projectUniqueName, vsSolution2, token);
+                        _projectSystemCache.AddProjectRestoreInfoSource(projectNames, projectRestoreInfoSource);
+                        taskCompletionSource.SetResult(true);
+                    }
+                    catch (OperationCanceledException oce)
+                    {
+                        taskCompletionSource.SetException(oce);
+                    }
+                    catch (Exception ex)
+                    {
+                        bool solutionFullyLoaded = false;
+                        try
+                        {
+                            solutionFullyLoaded = await IsSolutionFullyLoadedAsync(vsSolution2);
+                        }
+                        catch
+                        {
+                            // Do nothing.
+                        }
+
+                        if (solutionFullyLoaded)
+                        {
+                            taskCompletionSource.SetException(ex);
+                        }
+                        else
+                        {
+                            failedInits.Add(source);
+                        }
+                    }
+                }
+
+                _projectSystemCache.AddProjectRestoreInfoSource(projectNames, projectRestoreInfoSource);
```

Comment:
❗ This line is getting executed in two cases where it should not be allowed:

1. When OperationCanceledExecption is thrown by `FromIVsSolution2`
2. When a different exception is thrown, but the solution is not yet fully loaded

In each of these cases, `AddProjectRestoreInfoSource` throws `ArgumentNullException` for `projectNames`.

Reply:
@nkolev92 issue tracking is disabled for this repository, so I cannot convert this comment to an issue. This appears to be severely impacting our integration tests.

Reply:
NuGet/Home is the repo for issues.